buildflow/core/runtime/actors/process_pool.py

```diff
@@ -0,0 +1,118 @@
+import asyncio
+import random
+from typing import Dict
+
+import ray
+from ray.util.metrics import Gauge
+
+from buildflow import utils
+from buildflow.api import RuntimeAPI, RuntimeSnapshot, RuntimeStatus
+from buildflow.core.runtime.actors.pull_process_push import PullProcessPushActor  # noqa
+from buildflow.core.processor.base import Processor
+from buildflow.core.runtime.config import RuntimeConfig
+# TODO: add ability to load from env vars so we can set num_cpus
+# from buildflow.utils import load_config_from_env
+
+
+@ray.remote(num_cpus=0.1)
+class ProcessorReplicaPoolActor(RuntimeAPI):
+    """
+    This actor acts as a proxy reference for a group of replica Processors.
+    Runtime methods are forwarded to the replicas (i.e. "drain"). Includes
+    methods for adding and removing replicas (for autoscaling).
+    """
+
+    def __init__(self, processor: Processor, config: RuntimeConfig) -> None:
+        # configuration
+        self.processor = processor
+        self.num_threads = config.num_threads_per_process
+        # initial runtime state
+        self.replicas: Dict[str, PullProcessPushActor] = {}
+        self._status = RuntimeStatus.IDLE
+        # metrics
+        self.num_replicas_gauge = Gauge(
+            "processor_pool_num_replicas",
+            description="Current number of replicas. Goes up and down.",
+            tag_keys=("processor_name", ),
+        )
+        self.num_replicas_gauge.set_default_tags(
+            {"processor_name": self.processor.name})
+
+    def add_replicas(self, num_replicas: int):
+        if self._status != RuntimeStatus.RUNNING:
+            raise RuntimeError(
+                'Can only replicas to a processor pool that is running.')
+        for _ in range(num_replicas):
+            replica_id = utils.uuid()
+            self.replicas[replica_id] = PullProcessPushActor.remote(
+                self.processor)
+            if self._status == RuntimeStatus.RUNNING:
+                for _ in range(self.num_threads):
+                    self.replicas[replica_id].run.remote()
+
+        self.num_replicas_gauge.set(len(self.replicas))
+
+    async def remove_replicas(self, num_replicas: int):
+        if len(self.replicas) < num_replicas:
+            raise ValueError(
+                f'Cannot remove {num_replicas} replicas from '
+                f'{self.processor.name}. Only {len(self.replicas)} replicas '
+                'exist.')
+
+        actors_to_kill = []
+        actor_drain_tasks = []
+        # TODO: (maybe) Dont just kill randomly
+        replicas_to_remove = random.sample(self.replicas.keys(), num_replicas)
```

Comment:
I _think_ it is better to remove from the end, then it's more likely you'll remove all replicas that are on a node, and the node can get shutdown.

Ideally though we would probably look at what node an actor was running on and try to kill all actors on that node first.

Reply:
we might be able to use this to get the node: https://docs.ray.io/en/latest/ray-core/api/doc/ray.runtime_context.RuntimeContext.html#ray-runtime-context-runtimecontext but there's also an issue to handle this seperately so we can look into that later